File: amqp/src/main/scala/org/totalgrid/reef/messaging/javaclient/Session.scala
/**
 * Copyright 2011 Green Energy Corp.
 *
 * Licensed to Green Energy Corp (www.greenenergycorp.com) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  Green Energy Corp licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.totalgrid.reef.messaging.javaclient

import org.totalgrid.reef.api._
import org.totalgrid.reef.api.javaclient.{ IPromise, IResponse, ISubscription, ISession, IEventAcceptor }
import scalaclient.{ ClientSession, IPromise => IScalaPromise, Response => ScalaResponse }

/**
 * wraps a ProtoClient with some java helpers to convert to and from java lists
 */
class Session(val client: ClientSession) extends ISession {

  private implicit def convert[A](promise: IScalaPromise[ScalaResponse[A]]): IPromise[IResponse[A]] = new Promise[A](promise)

  final override def get[A](request: A): IPromise[IResponse[A]] = client.get(request)
  final override def delete[A](request: A): IPromise[IResponse[A]] = client.delete(request)
  final override def post[A](request: A): IPromise[IResponse[A]] = client.post(request)
  final override def put[A](request: A): IPromise[IResponse[A]] = client.put(request)

  private implicit def convert[A](sub: ISubscription[A]): RequestEnv = {
    val headers = new ServiceHandlerHeaders(new RequestEnv)
    headers.setSubscribeQueue(sub.getId)
    headers.env
  }

  final override def get[A](request: A, hdr: ISubscription[A]): IPromise[IResponse[A]] = client.get(request, hdr)
  final override def delete[A](request: A, hdr: ISubscription[A]): IPromise[IResponse[A]] = client.delete(request, hdr)
  final override def put[A](request: A, hdr: ISubscription[A]): IPromise[IResponse[A]] = client.put(request, hdr)
  final override def post[A](request: A, hdr: ISubscription[A]): IPromise[IResponse[A]] = client.post(request, hdr)

  final override def addSubscription[A](pd: ITypeDescriptor[A]): ISubscription[A] = {
    val sub = client.addSubscription[A](pd.getKlass)
    new SubscriptionWrapper(sub)
  }

  final override def addSubscription[A](pd: ITypeDescriptor[A], ea: IEventAcceptor[A]): ISubscription[A] = {
    val sub = client.addSubscription[A](pd.getKlass)
    val wrapped = new SubscriptionWrapper(sub)
    wrapped.start(ea)
    wrapped
  }

  final override def getDefaultEnv = new ServiceHandlerHeaders(client.getDefaultHeaders)

  final override def close() = client.close()
}

File: client-api/src/main/java/org/totalgrid/reef/api/japi/client/Session.java
/**
 * Copyright 2011 Green Energy Corp.
 * 
 * Licensed to Green Energy Corp (www.greenenergycorp.com) under one or more
 * contributor license agreements. See the NOTICE file distributed with this
 * work for additional information regarding copyright ownership. Green Energy
 * Corp licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.totalgrid.reef.api.japi.client;


import org.totalgrid.reef.api.japi.ReefServiceException;
import org.totalgrid.reef.api.japi.ServiceIOException;
import org.totalgrid.reef.api.japi.TypeDescriptor;

/**
 *  Session are provides access to REST service calls and subscriptions
 */
public interface Session
{

    /**
     * Make a service request using 'GET'
     *
     * @param request The request value to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> get( A request ) throws ReefServiceException;

    /**
     * Make a service request using 'DELETE'
     *
     * @param request The request value to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> delete( A request ) throws ReefServiceException;

    /**
     * Make a service request using 'POST'
     *
     * @param request The request value to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> post( A request ) throws ReefServiceException;

    /**
     * Make a service request using 'PUT'
     *
     * @param request The request value to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> put( A request ) throws ReefServiceException;

    /**
     * Make a service request using 'GET' that establishes a subscription
     *
     * @param request The request value to use
     * @param subscription Subscription object to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> get( A request, Subscription<A> subscription ) throws ReefServiceException;

    /**
     * Make a service request using 'DELETE' that establishes a subscription
     *
     * @param request The request value to use
     * @param subscription Subscription object to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> delete( A request, Subscription<A> subscription ) throws ReefServiceException;

    /**
     * Make a service request using 'POST' that establishes a subscription
     *
     * @param request The request value to use
     * @param subscription Subscription object to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> post( A request, Subscription<A> subscription ) throws ReefServiceException;

    /**
     * Make a service request using 'PUT' that establishes a subscription
     *
     * @param request The request value to use
     * @param subscription Subscription object to use
     * @param <A> The type of the request/response
     * @return A promise to the Response<A>
     * @throws ReefServiceException if the service request cannot be made
     */
    <A> Promise<Response<A>> put( A request, Subscription<A> subscription ) throws ReefServiceException;


    /**
     * Creates a subscription object that can be used during a request. After a successful request, the subscription can be started.
     *
     * @param descriptor TypeDescriptor corresponding to the type of the Subscription
     * @param <A> Type of the subscription
     * @return A Subscription object to use in a subsequent request
     * @throws ServiceIOException If the broker is not available, the descriptor is not recognized
     */
    <A> Subscription<A> addSubscription( TypeDescriptor<A> descriptor ) throws ServiceIOException;

    @Deprecated
    <A> Subscription<A> addSubscription( TypeDescriptor<A> descriptor, SubscriptionEventAcceptor<A> acceptor ) throws ServiceIOException;

    /**
     * Returns the immutable RequestHeaders
     * @return A reference to the immutable RequestHeaders object currently being used by the client.
     */
    RequestHeaders getHeaders();

    /**
     * Changes the
     * @param headers New RequestHeaders object to use as a default
     */
    void setHeaders( RequestHeaders headers );

    /**
     * Close the session, releasing the underlying resource. The session is dead and can never be used again.
     */
    void close();

}
